pep-0635.rst

```diff
@@ -0,0 +1,689 @@
+programming languages, its syntax slightly evolved and expanded.  The two
+first cases in the ``fib`` example above could be written more succinctly
+as ``case 0 | 1:`` with ``|`` denoting alternative patterns.  Moreover, the
+underscore ``_`` was generally accepted as a wildcard, a filler where neither
+the structure nor value of parts of a pattern were of substance.  Since the
+underscore is already frequently used in equivalent capacity in Python's
+iterable unpacking (e.g., ``_, _, third, _* = something``) we kept these
+universal standards.
+
+It is noteworthy that the concept of pattern matching has always been
+closely linked to the concept of functions.  The different case clauses
+have always been considered as something like semi-indepedent functions
+where pattern variables take on the role of parameters.  This becomes
+most apparent when pattern matching is written as an overloaded function,
+along the lines of::
+
+  def fib( 0 | 1 ):
+      return 1
+  def fib( n ):
+      return fib(n-1) + fib(n-2)
+
+Even though such a strict separation of case clauses into independent
+functions does not make sense in Python, we find that patterns share many
+syntactic rules with parameters, such as binding arguments to local
+variables only or that variable/parameter names must not be repeated for
+a particular pattern/function.
+
+With its emphasis on abstraction and encapsulation, object-oriented
+programming posed a serious challenge to pattern matching.  In short: in
+object-oriented programming, we can no longer view objects as tagged tuples.
+The arguments passed into the constructor do not necessarily specify the
+attributes or fields of the objects.  Moreover, there is no longer a strict
+ordering of an object's fields and some of the fields might be private and
+thus inaccessible.  And on top of this, the given object might actually be
+an instance of a subclass with slightly different structure.
+
+To address this challenge, patterns became increasingly independent of the
+original tuple constructors.  In a pattern like ``Node(left, right)``,
+``Node`` is no longer a passive tag, but rather a function that can actively
+check for any given object whether it has the right structure and extract a
+``left`` and ``right`` field.  In other words: the ``Node``-tag becomes a
+function that transforms an object into a tuple or returns ``None`` to
+indicate that it is not possible.
+
+In Python, we simply use ``isinstance()`` together with the ``__match_args__``
+field of a class to check whether an object has the correct structure and
+then transform some of its attributes into a tuple.  For the `Node` example
+above, for instance, we would have ``__match_args__ = ('left', 'right')`` to
+indicate that these two attributes should be extracted to form the tuple.
+That is, ``case Node(x, y)`` would first check whether a given object is an
+instance of ``Node`` and then assign ``left`` to ``x`` and ``right`` to ``y``,
+respectively.
```

Comment:
Doesn't this belong in a section about the class pattern (probably in the spec)? Presumably it's already described there. Maybe something not quite as nuts-and-bolts can be put here?

Reply:
The idea here was to complete the arch and circle back to how our Python implementation picks up these ideas of pattern matching out there.  So, yes, we can certainly add this in other places, but my intention here was to make the connection actually explicit.  But we could probably shorten it and leave the example out.